src/hammer-vlsi/hammer_vlsi/hammer_submit_command.py

```diff
@@ -0,0 +1,188 @@
+                subprocess_logger.debug(line.rstrip())
+                output_buf += line
+            else:
+                break
+        # TODO: check errors
+
+        return output_buf
+
+    def read_settings(self, tool_namespace: str, database: HammerDatabase) -> None:
+        assert("Should never get here")
+
+class HammerLSFSubmitCommand(HammerSubmitCommand):
+
+    @property
+    def bsub_binary(self) -> str:
+        """ Get the LSF bsub binary location """
+        try:
+            return self._bsub_binary
+        except AttributeError:
+            raise ValueError("LSF bsub binary location not set")
+
+    @bsub_binary.setter
+    def bsub_binary(self, value: str) -> None:
+        """ Set the LSF bsub binary location """
+        self._bsub_binary = value
+
+    @property
+    def num_cpus(self) -> int:
+        """ Get the LSF queue to use """
+        try:
+            return self._num_cpus
+        except AttributeError:
+            raise ValueError("Did not set the number of CPUs to use")
+
+    @num_cpus.setter
+    def num_cpus(self, value: int) -> None:
+        """ Set the LSF queue to use """
+        self._num_cpus = value # type: int
+
+    @property
+    def queue(self) -> Optional[str]:
+        """ Get the LSF queue to use """
+        try:
+            return self._queue
+        except AttributeError:
+            return None # use the default queue
+
+    @queue.setter
+    def queue(self, value: str) -> None:
+        """ Set the LSF queue to use """
+        self._queue = value # type: str
+
+    @property
+    def extra_args(self) -> List[str]:
+        """ Get the extra LSF args to use """
+        try:
+            return self._extra_args
+        except AttributeError:
+            # Use no extra args if empty
+            return [] # type : List[str]
+
+    @extra_args.setter
+    def extra_args(self, value: List[str]) -> None:
+        """ Set the extra LSF args to use """
+        self._extra_args = value
+
+    def read_settings(self, tool_namespace: str, database: HammerDatabase) -> None:
+        """
+        Read in LSF settings
+
+        TODO
+        """
+        pass
+
+
+    def bsub_args(self) -> List[str]:
+        args = [self.bsub_binary, "-K", "-n", "%d" % self.num_cpus] # always use -K to block
+        args.extend(self.extra_args)
+        if self.queue is not None:
+            args.extend(["-q", self.queue])
+        # TODO add more options as APIs (num_cpus, etc)
+        return args
+
+    def submit(self, args: List[str], env: Dict[str,str], logger: HammerVLSILoggingContext, cwd: str = None) -> str:
+        # TODO how to grab stdout/stderr from the LSF process without using temporary files?
+        logger.debug("Executing subprocess: " + ' '.join(self.bsub_args()) + '"' + ' '.join(args) + '"')
+        proc = subprocess.Popen(self.bsub_args() + [' '.join(args)], shell=False, stderr=subprocess.STDOUT, stdout=subprocess.PIPE, env=env, cwd=cwd)
+        atexit.register(proc.kill)
```

Comment:
Does this actually kill things? You don't need a `bkill` because you used the blocking flag?

Reply:
A `ctrl+C` to a `bsub` command with `-K` will kill it, so I assume this works.